### PR TITLE
Entry database update

### PR DESCRIPTION
**Pull request by ciliakas**
Refactored Entry Database to be instance based instead of static and changed all of front end to use Entry Database instead of Data Manager ( for now at least, needed to test how everything works, if I decide to refactor data manager as well, then I'll reintegrate Data Manager)
Changes are pretty similar to the previous User database update, the old Entry Database class was renamed as EntryDatabaseOld (it's still in use by Data Manager).

### CODE REVIEW

File: src/DataBase/EntryDatabase.cs [hunk 1]
```diff
@@ -139,69 +140,86 @@ public static bool Delete(Func<IEntryModel, bool> filter, EntryType entryType)
...
             {
-                var entriesToRemove = idArray.Select(id => db.Incomes.FirstOrDefault(x => x.Id == id && x.UserId == userId)).ToList();
-                db.Incomes.RemoveRange(entriesToRemove);
+                var list = new List<IncomeModel>();
+                foreach (var id in idArray)
+                {
+                    var temp = await Database.Incomes.FirstOrDefaultAsync(x => x.Id == id && x.UserId == userId);
+                    list.Add(temp);
+                }
+                Database.Incomes.RemoveRange(list);
```
Comment (Stokalas) [inline src/DataBase/EntryDatabase.cs]: Why was this changed from LINQ to foreach?

File: src/DataBase/EntryDatabase.cs [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
 using ePiggyWeb.DataManagement.Entries;
 using ePiggyWeb.Utilities;
+using Microsoft.EntityFrameworkCore;
 
 namespace ePiggyWeb.DataBase
 {
-    public static class EntryDatabase
+    public class EntryDatabase
     {
-        /*
```
Comment (Stokalas) [inline src/DataBase/EntryDatabase.cs:13]: So in short this file was changed from static to use async/await?

File: src/DataBase/EntryDatabaseOld.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using ePiggyWeb.DataBase.Models;
+using ePiggyWeb.DataManagement;
+using ePiggyWeb.DataManagement.Entries;
+using ePiggyWeb.Utilities;
+
+namespace ePiggyWeb.DataBase
+{
+    public static class EntryDatabaseOld
```
Comment (Stokalas) [inline src/DataBase/EntryDatabaseOld.cs:11]: Won't review this since, as I understand it is just to save old code.

File: src/Pages/EditEntry.cshtml.cs [hunk 4]
```diff
@@ -22,16 +23,20 @@ public class EditEntryModel : PageModel
...
         {
             var userId = int.Parse(User.FindFirst(ClaimTypes.Name).Value);
-            var dataManager = new DataManager(userId);
             EntryTypeInt = entryType;
 
             Entry = entryType == 1
-                ? (Entry)dataManager.Income.EntryList.FirstOrDefault(x => x.Id == id)
-                : (Entry)dataManager.Expenses.EntryList.FirstOrDefault(x => x.Id == id);
+                ? (Entry) await EntryDatabase.ReadAsync(id, userId, EntryType.Income)
+                : (Entry) await EntryDatabase.ReadAsync(id, userId, EntryType.Expense);
```
Comment (Stokalas) [inline src/Pages/EditEntry.cshtml.cs:39]: Seems much cleaner now, eventhough it's testing file :D

File: src/Pages/ForgotPassword.cshtml.cs [hunk 5]
```diff
@@ -36,10 +36,10 @@ public class ForgotPasswordModel : PageModel
 
         private EmailSender EmailSender { get; }
         private UserDatabase UserDatabase { get; }
-        public ForgotPasswordModel(EmailSender emailSender, UserDatabase userDatabase)
+        public ForgotPasswordModel(UserDatabase userDatabase, EmailSender emailSender)
```
Comment (Stokalas) [inline src/Pages/ForgotPassword.cshtml.cs]: Just interested, why was this changed?

File: src/Startup.cs [hunk 6]
```diff
@@ -42,6 +42,11 @@ public void ConfigureServices(IServiceCollection services)
                 var db = provider.GetService<PiggyDbContext>();
                 return new UserDatabase(db);
             });
+            services.AddScoped(provider =>
+            {
+                var db = provider.GetService<PiggyDbContext>();
+                return new EntryDatabase(db);
+            });
             services.AddScoped<EmailSender>();
```
Comment (Stokalas) [inline src/Startup.cs]: As always, nice to see DI happening

**Review by Stokalas — APPROVED** [review]
Seems to be well done, left few questions, as always.

File: src/DataBase/EntryDatabase.cs (hunk 2 above)
Comment (ciliakas) [reply src/DataBase/EntryDatabase.cs:13]: Yeah, in essence

File: src/DataBase/EntryDatabase.cs (hunk 1 above)
Comment (ciliakas) [reply src/DataBase/EntryDatabase.cs]: Wanted to write this as quick possible, and wasn't sure how to change this expression to be async. Spent some time and figured out how to write as a linq expression, so this was a good question to ask :D

File: src/Pages/ForgotPassword.cshtml.cs (hunk 5 above)
Comment (ciliakas) [reply src/Pages/ForgotPassword.cshtml.cs]: No reason really I think, just being ocd about pointless things :D

File: src/DataBase/EntryDatabase.cs [hunk 7]
```diff
@@ -142,49 +142,30 @@ public async Task<bool> DeleteAsync(Expression<Func<IEntryModel, bool>> filter,
...
-            else
-            {
-                var entriesToRemove = await Database.Expenses.Where(filter).Cast<ExpenseModel>().ToListAsync();
-                Database.Expenses.RemoveRange(entriesToRemove);
-            }
-
+                EntryType.Income => await Database.Incomes.Where(filter).Cast<IncomeModel>().ToListAsync(),
+                _ => await Database.Expenses.Where(filter).Cast<ExpenseModel>().ToListAsync()
+            };
+            Database.RemoveRange(entriesToRemove);
```
Comment (Stokalas) [inline src/DataBase/EntryDatabase.cs]: How does this work, or source from where to read about?

File: src/DataBase/EntryDatabase.cs [hunk 8]
```diff
@@ -142,49 +142,30 @@ public async Task<bool> DeleteAsync(Expression<Func<IEntryModel, bool>> filter,
...
-                }
-                Database.Expenses.RemoveRange(list);
-            }
-
+                EntryType.Income => await Task.WhenAll(idArray.Select(selector: async id =>
+                    await Database.Incomes.FirstOrDefaultAsync(x => x.Id == id && x.UserId == userId))),
+                _ => await Task.WhenAll(idArray.Select(selector: async id =>
+                    await Database.Expenses.FirstOrDefaultAsync(x => x.Id == id && x.UserId == userId)))
+            };
+            Database.RemoveRange(entriesToRemove);
```
Comment (Stokalas) [inline src/DataBase/EntryDatabase.cs]: Same here

**Review by Stokalas — APPROVED** [review]
Nice to see, that you managed to keep Linq with async, just a few questions. Approving

File: src/DataBase/EntryDatabase.cs (hunk 7 above)
Comment (ciliakas) [reply src/DataBase/EntryDatabase.cs]: I think line is more clear when I removed the (as I understand) redundant cast

     await Database.Incomes.Where(filter).ToListAsync()


Source to read up on: https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.queryableextensions.tolistasync?view=entity-framework-6.2.0

File: src/DataBase/EntryDatabase.cs (hunk 8 above)
Comment (ciliakas) [reply src/DataBase/EntryDatabase.cs]: Thank you for asking about this, because this led me to some big improvements for this part :D Before this was generating threading issues, I think it's because of the two await statements, but I'm not completely sure why. 
Now the code looks like this:

            var filter = PredicateBuilder.BuildEntryFilter(idArray, userId);
            return await DeleteListAsync(filter, entryType);

You should refer to the new pull request about how this works, basically it creates a single predicate and does a single search, so not only does this work now ( tested it) also it should work faster, because of a query for every id in the list, now it just uses one query for all of them

File: src/DataManagement/Dm.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+using ePiggyWeb.DataManagement.Goals;
+using ePiggyWeb.Utilities;
+
+namespace ePiggyWeb.DataManagement
+{
+    public class Dm
+    {
+        public IEntryManager Income { get; }
+        public IEntryManager Expenses { get; }
+        public IGoalManager Goals { get; }
```
Comment (Stokalas) [inline src/DataManagement/Dm.cs]: What is this class?

Comment (ciliakas) [reply src/DataManagement/Dm.cs]: This somehow got here by accident, was supposed to be stashed away :D It's data manager rework I had started working on

**Review by Stokalas — APPROVED** [review]
Seems ok now
